BlueprintUI/Sources/Layout/Stack.swift

```diff
@@ -368,49 +407,86 @@ extension StackLayout {
             totalPriority = 1
         }
 
-        for index in 0..<basisSizes.count {
+        let axisSegments = zip(basisSizes, growPriorities).map { (basis, growPriority) -> Segment in
+            let sizeAdjustment = (growPriority / totalPriority) * extraSize
+            let origin = axisOrigin
+            let magnitude = basis + sizeAdjustment
 
-            frames[index].origin.axis = axisOrigin
+            axisOrigin = origin + magnitude + space
 
-            let basis = basisSizes[index]
+            return Segment(origin: origin, magnitude: magnitude)
+        }
 
-            frames[index].size.axis = basis.axis + ((growPriorities[index] / totalPriority) * extraSize)
+        return axisSegments
+    }
 
-            axisOrigin = frames[index].maxAxis + space
+    private func _crossSegments(
+        for items: [Measurable],
+        axisConstraints: [CGFloat],
+        crossConstraint: SizeConstraint.Axis
+    ) -> [Segment] {
+        // First, measure cross magnitudes based on axis constraints
+        let crossMagnitudes = zip(items, axisConstraints).map { (item, axisConstraint) -> CGFloat in
+            let vector = VectorConstraint(axis: .atMost(axisConstraint), cross: crossConstraint)
+            let constraint = vector.constraint(axis: axis)
+            let measuredSize = item.measure(in: constraint)
+
+            return measuredSize.cross(on: axis)
         }
 
-        return frames
-    }
+        // Then pick the max cross value based on the constraint
+        let maxCross: CGFloat
+        switch crossConstraint {
+        case .unconstrained:
+            maxCross = crossMagnitudes.reduce(0, max)
+        case .atMost(let max):
+            maxCross = max
+        }
+
+        // Finally, form segments from the magnitudes and the alignment option
+        let segments = zip(items, crossMagnitudes).map { (item, measuredCross) -> Segment in
+            let origin: CGFloat
+            let magnitude: CGFloat
 
-    fileprivate func _calculateCross(basisSizes: [Vector], layoutSize: Vector) -> [Frame] {
-        return basisSizes.map { (measuredSize) -> Frame in
-            var result = Frame.zero
             switch alignment {
             case .center:
-                result.origin.cross = (layoutSize.cross - measuredSize.cross) / 2.0
-                result.size.cross = measuredSize.cross
+                origin = (maxCross - measuredCross) / 2.0
```

Comment:
It didnt have it before, but should this be rounded?

Reply:
I don't think we have enough context to round correctly here. I'm hoping #64 will solve this generally.